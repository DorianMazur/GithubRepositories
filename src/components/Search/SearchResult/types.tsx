export interface IDefaultProps {
    open: boolean | undefined;
}

export interface IDefaultState {
    data: null | { avatar: string; bio: string; repos: Array<{ name: string; stargazers_count: number; url: string }> };
    loading: boolean;
}

export interface IProps extends Partial<IDefaultProps> {
    onClose: () => void;
    searchResult: string | undefined;
}

export interface IState extends Partial<IDefaultState> {
}
