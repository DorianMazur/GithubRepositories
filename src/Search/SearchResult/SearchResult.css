.searchResult{
 text-align: center;
}

.searchResult__avatar{
    margin: 6px auto;
}

.searchResult__container{
    padding:10px;
}

.searchResult__link{
    display: block;
}