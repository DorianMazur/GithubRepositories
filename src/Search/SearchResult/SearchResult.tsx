import React, { Component } from 'react';
import LinearProgress from '@material-ui/core/LinearProgress';
import Avatar from '@material-ui/core/Avatar';
import DialogTitle from '@material-ui/core/DialogTitle';
import Dialog from '@material-ui/core/Dialog';
import Typography from '@material-ui/core/Typography';
import Link from '@material-ui/core/Link';
import StarIcon from '@material-ui/icons/Star';
import { getUserInfo, getRepositories } from '../../Api/api';
import { orderBy } from 'lodash';
import './SearchResult.css';

type SearchResultProps = {
    open: boolean;
    onClose: () => void;
    searchResult: string | null;
};

type SearchResultStat = {
    data: null | { avatar: string; bio: string; repos: Array<{ name: string; stargazers_count: number; url: string }> };
    loading: boolean;
};

export default class SearchResult extends Component<SearchResultProps, SearchResultStat> {
    constructor(props: SearchResultProps) {
        super(props);
        this.state = { loading: true, data: null };
        this.fetchData = this.fetchData.bind(this);
    }

    fetchData(): void {
        const { searchResult } = this.props;
        if (!searchResult) return;
        const calls = [getUserInfo(searchResult), getRepositories(searchResult)];
        Promise.all(calls).then(responses => {
            const mostPopularRepos = orderBy(responses[1].response, ['stargazers_count', 'watchers_count']);
            const data = {
                avatar: responses[0].response.avatar_url,
                bio: responses[0].response.bio,
                repos: mostPopularRepos.slice(0, 3),
            };
            this.setState({ loading: false, data });
        });
    }

    render(): JSX.Element | null {
        const { open, onClose, searchResult } = this.props;
        const { loading, data } = this.state;
        if (!searchResult) return null;
        return (
            <Dialog
                onClose={onClose}
                onEnter={this.fetchData}
                aria-labelledby="simple-dialog-title"
                open={open}
                className="searchResult"
            >
                <DialogTitle id="simple-dialog-title">Acount name: {searchResult}</DialogTitle>
                {!loading && data ? (
                    <div className="searchResult__container">
                        <Avatar alt={searchResult} src={data.avatar} className="searchResult__avatar" />
                        <Typography variant="subtitle1">{data.bio || 'No Bio at all'}</Typography>
                        {data.repos.length ? (
                            <div>
                                <Typography variant="h6">Most popular repositories:</Typography>
                                {data.repos.map(repo => (
                                    <Link href={repo.url} key={repo.name} className="searchResult__link">
                                        {repo.name} <StarIcon /> {repo.stargazers_count}
                                    </Link>
                                ))}
                            </div>
                        ) : null}
                    </div>
                ) : (
                    <LinearProgress />
                )}
            </Dialog>
        );
    }
}
