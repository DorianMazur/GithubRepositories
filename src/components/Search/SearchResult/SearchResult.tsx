import * as React from 'react';
import LinearProgress from '@material-ui/core/LinearProgress';
import Avatar from '@material-ui/core/Avatar';
import DialogTitle from '@material-ui/core/DialogTitle';
import Dialog from '@material-ui/core/Dialog';
import Typography from '@material-ui/core/Typography';
import Link from '@material-ui/core/Link';
import StarIcon from '@material-ui/icons/Star';
import { getUserInfo, getRepositories } from '../../../api/api';
import { orderBy } from 'lodash';
import { IState, IProps, IDefaultProps, IDefaultState } from './types';
import './SearchResult.css';

export default class SearchResult extends React.Component<IProps, IState> {
    static defaultProps: IDefaultProps = {
        open: false,
    };

    readonly state: IDefaultState = {
        loading: true,
        data: null,
    };

    fetchData = () => {
        const { searchResult } = this.props;
        if (!searchResult) return;
        const calls = [getUserInfo(searchResult), getRepositories(searchResult)];
        Promise.all(calls).then(responses => {
            const mostPopularRepos = orderBy(responses[1].response, ['stargazers_count', 'watchers_count']);
            const data = {
                avatar: responses[0].response.avatar_url,
                bio: responses[0].response.bio,
                repos: mostPopularRepos.slice(0, 3),
            };
            this.setState({ loading: false, data });
        });
    };

    onClose = () => {
        this.setState({ loading: true, data: null }, () => this.props.onClose());
    };

    render(): JSX.Element | null {
        const { open, searchResult } = this.props;
        const { loading, data } = this.state;
        if (!searchResult) return null;
        return (
            <Dialog
                onClose={this.onClose}
                onEnter={this.fetchData}
                aria-labelledby="simple-dialog-title"
                open={open || false}
                className="searchResult"
            >
                <DialogTitle id="simple-dialog-title">Acount name: {searchResult}</DialogTitle>
                {!loading && data ? (
                    <div className="searchResult__container">
                        <Avatar alt={searchResult} src={data.avatar} className="searchResult__avatar" />
                        <Typography variant="subtitle1">{data.bio || 'No Bio at all'}</Typography>
                        {data.repos.length ? (
                            <div>
                                <Typography variant="h6">Most popular repositories:</Typography>
                                {data.repos.map(repo => (
                                    <Link href={repo.url} key={repo.name} className="searchResult__link">
                                        {repo.name} <StarIcon /> {repo.stargazers_count}
                                    </Link>
                                ))}
                            </div>
                        ) : null}
                    </div>
                ) : (
                    <LinearProgress />
                )}
            </Dialog>
        );
    }
}
