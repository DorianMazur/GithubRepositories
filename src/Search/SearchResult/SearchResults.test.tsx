import '@testing-library/jest-dom';
import React from 'react';

import { render } from '@testing-library/react';
import SearchResult from './SearchResult';

test('SearchResults render correctly closed with null searchResult', () => {
    const container = render(<SearchResult open={false} onClose={jest.fn()} searchResult={null} />);
    expect(container.baseElement).toMatchSnapshot();
});

test('SearchResults render correctly closed without null searchResult', () => {
    const container = render(<SearchResult open={false} onClose={jest.fn()} searchResult={'Test'} />);
    expect(container.baseElement).toMatchSnapshot();
});

test('SearchResults render correctly open with null searchResult', () => {
    const container = render(<SearchResult open={true} onClose={jest.fn()} searchResult={null} />);
    expect(container.baseElement).toMatchSnapshot();
});

test('SearchResults render correctly open without null searchResult', () => {
    const container = render(<SearchResult open={true} onClose={jest.fn()} searchResult={'Test'} />);
    expect(container.baseElement).toMatchSnapshot();
});
