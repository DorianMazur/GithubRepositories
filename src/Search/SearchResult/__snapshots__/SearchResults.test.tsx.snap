// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults render correctly closed with null searchResult 1`] = `
<body>
  <div />
</body>
`;

exports[`SearchResults render correctly closed without null searchResult 1`] = `
<body>
  <div />
</body>
`;

exports[`SearchResults render correctly open with null searchResult 1`] = `
<body>
  <div />
</body>
`;

exports[`SearchResults render correctly open without null searchResult 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root searchResult"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby="simple-dialog-title"
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
          id="simple-dialog-title"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            Acount name: 
            Test
          </h2>
        </div>
        <div
          class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate"
          role="progressbar"
        >
          <div
            class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
          />
          <div
            class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
          />
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
